### DISCUSSION TITLE
[Embeded mode]How to set the default task for loading scene?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1087] [Q&A] [answered]
Hi:

Each time when the game starts, we hope to load some resources which may take some times
so hope players could see the loading scene and the percentage
but couldn't find a way to set the default task for the loading scene
Any idea how to set the default task?

Thanks

**Comment by chengenzhao** [comment]
Hmm....
I think I find a way to do it
```java
    @Override
    public void onCreate() {

        var task = new Task<Void>(){
            @Override
            protected Void call() throws Exception {
                updateProgress(0,100);
                FXGL.image("player" + separator + "upper.png", 1400 /2, 350 /2);
                updateProgress(20,100);
                FXGL.image("player" + separator + "run.png", 2880/2, 220/2);
                updateProgress(40,100);
                FXGL.image("player" + separator + "jump.png", 2400/2, 220/2);
                updateProgress(60,100);
                FXGL.image("player" + separator + "fire.png", 4500/2, 250/2);
                updateProgress(80,100);
                return null;
            }
        };
        progressProp.bind(task.progressProperty());
        new Thread(task).start();
}
```
but not sure this is the right way to do it

**Reply by chengenzhao** [reply]
Hmmm... it is not working
The game wouldn't wait the task complete 
I tried bind(task) and pushNewTask(task)
but no matter how I set the task,
the game won't wait the task complete and directly jump to the next scene

**Comment by AlmasB** [comment]
Have a look at this [sample](https://github.com/AlmasB/FXGL/blob/a8598950d0fa18181efdc9ab63231c966edeb519/fxgl-samples/src/main/java/intermediate/CustomLoadingSample.java), perhaps something useful there?

**Reply by chengenzhao** [reply]
Hmm... not working
I got this error

> 18:36:24.012 [JavaFX Application Thread] WARN  EmbeddedPaneWindow   - Cannot set to Loading. Probably because subscenes are present.

```java
  @Override
  protected void initInput() {
    super.initInput();
    getGameScene().setCursor(Cursor.DEFAULT);

    var task = new Task<Void>() {
      @Override
      protected Void call() throws Exception {
        try {
          updateProgress(10,100);
          Thread.sleep(1000);
          FXGL.image("background" + separator + "background.png");
          updateProgress(20,100);
          Thread.sleep(1000);
          updateProgress(50,100);
          Thread.sleep(5000);
        } catch (InterruptedException e) {
          e.printStackTrace();
        }
        return null;
      }
    };
//    loading.bind(task);
    getGameController().gotoLoading(task);
  }
```

Without above code, the loading scene could be successfully set
but if add in above code, then the loading scene gets error

**Reply by chengenzhao** [reply]
Tried another project, also gets this error

> [JavaFX Application Thread] WARN  EmbeddedPaneWindow   - Cannot set to Loading. Probably because subscenes are present.

**Comment by AlmasB** [comment, marked as answer]
There is no need to explicitly call `loading()` since `initGame()` already runs on a background thread. While initGame() runs, FXGL will display the loading scene, which can be customized. So:

1. Using scene factory, provide your own implementation of LoadingScene.
2. Something like this:

```
public void initGame() {
                updateProgress(0,100);
                FXGL.image("player" + separator + "upper.png", 1400 /2, 350 /2);
                updateProgress(20,100);
                FXGL.image("player" + separator + "run.png", 2880/2, 220/2);
                updateProgress(40,100);
                FXGL.image("player" + separator + "jump.png", 2400/2, 220/2);
                updateProgress(60,100);
                FXGL.image("player" + separator + "fire.png", 4500/2, 250/2);
                updateProgress(80,100);
}
```

**Reply by chengenzhao** [reply]
Got it
it works
But I have to set property of game
and use it in the initGame method
like this
```java
public class Game extends GameApplication {
    Loading loading;
    
    @Override
    protected void initSettings(GameSettings settings) {
        settings.setSceneFactory(new SceneFactory(){

            @Override
            public LoadingScene newLoadingScene() {
                loading = new Loading();
                return loading;
            }
        });

...

//use it

    @Override
    protected void initGame(){
        super.initGame();
        FXGL.image("player" + separator + "upper.png", 1400 / 2, 350 / 2);
        loading.setProgress(0.2);
        FXGL.image("player" + separator + "run.png", 2880 / 2, 220 / 2);
        loading.setProgress(0.3);
        FXGL.image("player" + separator + "jump.png", 2400 / 2, 220 / 2);
        loading.setProgress(0.5);
        FXGL.image("player" + separator + "fire.png", 4500 / 2, 250 / 2);
        
```

**Reply by AlmasB** [reply]
I assume you are then handling `loading.setProgress` on the JavaFX thread to avoid concurrency issues since initGame() is on a bg thread, but yes the above is fine.

**Reply by AlmasB** [reply]
Also, as a side note, `super.initGame()` is not needed. None of the overridden methods require calling `super.*`.

**Reply by chengenzhao** [reply]
> I assume you are then handling `loading.setProgress` on the JavaFX thread to avoid concurrency issues since initGame() is on a bg thread, but yes the above is fine.

Thanks 
Will use it in 
```java
Platform.runLater(()->progressProp.setValue(progress));
```
